rs/bitcoin/kyt/src/lib.rs

```diff
@@ -92,33 +75,114 @@ async fn get_tx(tx_id: String) -> Result<Transaction, BitcoinTxError> {
         }),
         headers: request_headers,
     };
-    let cycles = 49_140_000 + 1024 * 5_200 + 10_400 * 400 * 1024; // 1 KiB request, 400 KiB response
+    let cycles = get_tx_cycle_cost(buffer_size);
     match http_request(request, cycles).await {
         Ok((response,)) => {
             // TODO(XC-158): ensure response is 200 before decoding
             let tx = Transaction::consensus_decode(&mut response.body.as_slice())
-                .map_err(BitcoinTxError::Encoding)?;
+                .map_err(|err| GetTxError::Encoding(err.to_string()))?;
             // Verify the correctness of the transaction by recomputing the transaction ID.
-            let decoded_tx_id = tx.compute_txid().to_string();
-            if decoded_tx_id != tx_id {
-                return Err(BitcoinTxError::TxIdMismatch {
-                    expected: tx_id,
-                    decoded: decoded_tx_id,
+            let decoded_tx_id = tx.compute_txid();
+            if decoded_tx_id.as_ref() as &[u8; 32] != tx_id.as_ref() {
+                return Err(GetTxError::TxIdMismatch {
+                    expected: tx_id.into(),
+                    decoded: *decoded_tx_id.as_ref(),
                 });
             }
             Ok(tx)
         }
+        Err((r, m)) if is_response_too_large(&r, &m) => Err(GetTxError::ResponseTooLarge),
         Err((r, m)) => {
             // TODO(XC-158): maybe try other providers and also log the error.
             println!("The http_request resulted into error. RejectionCode: {r:?}, Error: {m}");
-            Err(BitcoinTxError::Rejected {
+            Err(GetTxError::Rejected {
                 code: r,
                 message: m,
             })
         }
     }
 }
 
+impl HasOutPoint for TxIn {
+    fn txid(&self) -> Txid {
+        Txid::from(*(self.previous_output.txid.as_ref() as &[u8; 32]))
+    }
+    fn vout(&self) -> u32 {
+        self.previous_output.vout
+    }
+}
+
+impl TransactionLike for Transaction {
+    type Input = TxIn;
+    type AddressError = GetTxError;
+
+    fn iter_inputs(&self) -> impl Iterator<Item = &TxIn> {
+        self.input.iter()
+    }
+    fn output_address(&self, vout: u32) -> Result<Address, GetTxError> {
+        let output = &self.output[vout as usize];
+        Address::from_script(&output.script_pubkey, Network::Bitcoin).map_err(GetTxError::Address)
+    }
+}
+
+struct KytCanisterState;
+
+impl FetchState<Transaction> for KytCanisterState {
```

Comment:
> Does that make sense?

Yes, I understand. From experience with ckETH what works well is to usually have the following pattern

1. top-level method in `main`, 
    1. Acquire guard
    2. access to thread-local to read/hold mutable reference to state.
2. Delegate to internal method with the reference. This internal method is a pure Rust method (no thread_local, no ic_cdk, etc.)

This allows then for the following testing:

1. test pure method with standard unit tests. Since most of the logic should be in this method, this is where the bulk of the test should be.
2. Test outer method via state machine /pocket IC tests, for cases not covered by unit tests.

I have the impression that this offers a clean separation between state and logic, which also allows for easier testing. Note that this is definitely not a must, just an observation from past experience, so it's also ok if you want to try out your way!